pennylane/ops/op_math/evolve.py

```diff
@@ -0,0 +1,186 @@
+    r"""Parametrized evolution gate.
+
+    For a time-dependent Hamiltonian of the form
+    :math:`H(v, t) = H_\text{drift} + \sum_j f_j(v, t) H_j` it implements the corresponding
+    time-evolution operator :math:`U(t_1, t_2)`, which is the solution to the time-dependent
+    Schrodinger equation
+
+    .. math:: \frac{d}{dt}U(t) = -i H(v, t) U(t).
+
+
+    Under the hood, it is using a numerical ordinary differential equation solver.
+
+    Args:
+        H (ParametrizedHamiltonian): hamiltonian to evolve
+        params (ndarray): trainable parameters
+        t (Union[float, List[float]]): If a float, it corresponds to the duration of the evolution.
+            If a list of two floats, it corresponds to the initial time and the final time of the
+            evolution. Note that such absolute times only have meaning within an instance of ``ParametrizedEvolution`` and will not affect other gates.
+        dt (float): the time step used by the differential equation solver to evolve the
+            time-dependent Hamiltonian. Defaults to XXX.
+        time (str, optional): The name of the time-based parameter in the parametrized Hamiltonian.
+            Defaults to "t".
+        do_queue (bool): determines if the scalar product operator will be queued. Default is True.
+        id (str or None): id for the scalar product operator. Default is None.
+    """
+
+    _name = "ParametrizedEvolution"
+    num_wires = AnyWires
+    # pylint: disable=too-many-arguments, super-init-not-called
+    def __init__(
+        self, H: ParametrizedHamiltonian, params: list, t, dt=None, time="t", do_queue=True, id=None
+    ):
+        self.H = H
+        self.time = time
+        self.dt = dt
+        self.h_params = params
+        self.t = [0, t] if qml.math.size(t) == 1 else t
+        super().__init__(wires=H.wires, do_queue=do_queue, id=id)
+
+    # pylint: disable=arguments-renamed, invalid-overridden-method
+    @property
+    def has_matrix(self):
+        return all(op.has_matrix or isinstance(op, qml.Hamiltonian) for op in self.H.ops)
+
+    # pylint: disable=import-outside-toplevel
+    def matrix(self, wire_order=None):
+        try:
+            import jax.numpy as jnp
+            from jax.experimental.ode import odeint
+        except ImportError as e:
+            raise ImportError(
+                "Module jax is required for ``ParametrizedEvolution`` class. "
+                "You can install jax via: pip install jax"
+            ) from e
+        y0 = jnp.eye(2 ** len(self.wires), dtype=complex)
+
+        def fun(y, t, params):
+            """dy/dt = -i H(t) y"""
+            kwargs = {self.time: t}
+            return -1j * qml.matrix(self.H(params, **kwargs)) @ y
+
+        if self.dt is None:
+            # TODO: Figure out what is 'p', and best values for 'rtol' and 'atol'
+            self.dt = guess_dt(fun, self.t[0], y0, 2, *self.h_params)
+
+        t = jnp.arange(self.t[0], self.t[1], self.dt, dtype=float)
```

Comment:
What's the difference? `self.t` should have length 2

Reply:
a user could pass a longer list of ts (thinking that this will be executed intermediate points, which is quite standard behavior for ode solvers. This way it is making sure it is taking only the first and last value. Alternative is to add a test that t is of size 2 (when it is not of size 1, see line 100)